openhcl/virt_mshv_vtl/src/processor/hardware_cvm/mod.rs

```diff
@@ -201,38 +201,104 @@ impl<T: CpuIo, B: HardwareIsolatedBacking> UhHypercallHandler<'_, '_, T, B> {
+            | HvX64RegisterName::Cr8
+            | HvX64RegisterName::Dr0
+            | HvX64RegisterName::Dr1
+            | HvX64RegisterName::Dr2
+            | HvX64RegisterName::Dr3
+            | HvX64RegisterName::Dr7
+            | HvX64RegisterName::Es
+            | HvX64RegisterName::Cs
+            | HvX64RegisterName::Ss
+            | HvX64RegisterName::Ds
+            | HvX64RegisterName::Fs
+            | HvX64RegisterName::Gs
+            | HvX64RegisterName::Tr
+            | HvX64RegisterName::Ldtr
+            | HvX64RegisterName::Gdtr
+            | HvX64RegisterName::Idtr
+            | HvX64RegisterName::Rip
+            | HvX64RegisterName::Rflags
+            | HvX64RegisterName::Rax
+            | HvX64RegisterName::Rcx
+            | HvX64RegisterName::Rdx
+            | HvX64RegisterName::Rbx
+            | HvX64RegisterName::Rsp
+            | HvX64RegisterName::Rbp
+            | HvX64RegisterName::Rsi
+            | HvX64RegisterName::Rdi
+            | HvX64RegisterName::R8
+            | HvX64RegisterName::R9
+            | HvX64RegisterName::R10
+            | HvX64RegisterName::R11
+            | HvX64RegisterName::R12
+            | HvX64RegisterName::R13
+            | HvX64RegisterName::R14
+            | HvX64RegisterName::R15
+            | HvX64RegisterName::Pat => {
+                // Architectural registers can only be accessed by a higher VTL.
+                if vtl >= self.intercepted_vtl {
+                    return Err(HvError::AccessDenied);
+                }
+                Ok(())
+            }
+            HvX64RegisterName::TscAux => {
+                // Architectural registers can only be accessed by a higher VTL.
+                if vtl >= self.intercepted_vtl {
+                    return Err(HvError::AccessDenied);
+                }
 
-        for (i, (&name, output)) in zip(registers, output).enumerate() {
-            *output = self.get_vp_register(name, vtl).map_err(|e| (e, i))?;
+                if self.vp.partition.caps.tsc_aux {
+                    Ok(())
+                } else {
+                    Err(HvError::InvalidParameter)
+                }
+            }
+            _ => Ok(()),
         }
-
-        Ok(())
     }
 
     fn get_vp_register(
         &mut self,
         name: hvdef::HvRegisterName,
         vtl: GuestVtl,
     ) -> HvResult<hvdef::HvRegisterValue> {
+        self.validate_register_access(vtl, name)?;
+        // TODO: when get vp register i.e. in access vp state gets refactored,
+        // clean this up.
+
+        let (registers, msrs, debug_regs, cache_control, xcr, tsc_aux) = {
+            let mut vp_state = self.vp.access_state(vtl.into());
+            (
+                vp_state.registers().unwrap(),
+                vp_state.virtual_msrs().unwrap(),
+                vp_state.debug_regs().unwrap(),
+                vp_state.cache_control().unwrap(),
+                vp_state.xcr().unwrap(),
+                vp_state.tsc_aux().unwrap(),
```

Comment:
these can be potentially expensive right? we probably want to only get the ones we have to

Reply:
Also should we handle errors instead of unwrapping?

Reply:
I think when I looked at these paths, I didn't expect there to be any actual errors, but I can check again.

Reply:
Well they must return something, hence the unwraps, right?

Reply:
I meant that they don't do any operations that can cause an error. E.g. for snp, for most of them we're just setting the vmsa (which I would not expect to fail) and returning Ok(())

Reply:
Right but what about TDX, or future backings?